File: results/2026-06-30_191904/tasks/12-pr-label-assigner/default-sonnet5/generated-code/run_act_tests.py
#!/usr/bin/env python3
"""
Outer test harness: actually executes the GitHub Actions workflow through
`act` (nektos/act) in Docker, rather than invoking label_assigner.py
directly. Every functional test case for this task is a *step* inside the
single `pr-label-assigner.yml` workflow (see the "Scenario - ..." steps), so
one `act push --rm` run exercises all of them in one pass -- this keeps us
well under the "at most 3 act push runs" budget while still asserting exact,
per-scenario expected values.

Steps:
  1. Build a throwaway git repo under a temp dir containing the project
     files + fixture data.
  2. Run `act push --rm`, capturing combined stdout/stderr.
  3. Append the captured output to act-result.txt (this repo's cwd), clearly
     delimited.
  4. Assert the process exited 0.
  5. Assert each expected exact label string appears (one per mocked PR
     scenario) and that both jobs report success.
"""
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent
ACT_RESULT_PATH = REPO_ROOT / "act-result.txt"

# Files/dirs copied into the throwaway git repo that `act` runs against.
PROJECT_ENTRIES = [
    "label_assigner.py",
    "label_rules.json",
    "fixtures",
    "tests",
    ".github",
    ".actrc",
]

# Exact expected values per mocked PR scenario -- these mirror the assertions
# already proven at the unit level in tests/test_label_assigner.py, now
# re-verified as they actually run inside the containerized workflow.
EXPECTED_SCENARIO_OUTPUT = {
    "docs-only PR": "Labels: documentation",
    "api change (priority conflict resolution)": "Labels: api, size/small",
    "mixed changes (multiple independent labels)": "Labels: documentation, size/large, tests",
    "no matching rules": "Labels: (none)",
}

EXPECTED_UNIT_TEST_SUMMARY = "23 passed"


def build_temp_repo() -> Path:
    tmp_dir = Path(tempfile.mkdtemp(prefix="pr-label-assigner-act-"))
    for entry in PROJECT_ENTRIES:
        src = REPO_ROOT / entry
        dst = tmp_dir / entry
        if src.is_dir():
            shutil.copytree(src, dst)
        else:
            shutil.copy2(src, dst)

    run(["git", "init", "-q"], cwd=tmp_dir)
    run(["git", "config", "user.email", "act-test@example.com"], cwd=tmp_dir)
    run(["git", "config", "user.name", "act-test"], cwd=tmp_dir)
    run(["git", "add", "-A"], cwd=tmp_dir)
    run(["git", "commit", "-q", "-m", "test commit for act"], cwd=tmp_dir)
    return tmp_dir


def run(cmd, cwd=None):
    result = subprocess.run(cmd, cwd=cwd, capture_output=True, text=True)
    if result.returncode != 0:
        raise RuntimeError(f"Command failed: {cmd}\nstdout:\n{result.stdout}\nstderr:\n{result.stderr}")
    return result


def run_act_push(repo_dir: Path):
    # --pull=false: the custom act-ubuntu-pwsh image is already built locally
    # (see Dockerfile.act); act's default --pull=true tries to re-pull it
    # from a registry and fails with a Docker Hub auth error.
    return subprocess.run(
        ["act", "push", "--rm", "--pull=false"],
        cwd=repo_dir, capture_output=True, text=True,
    )


def main() -> int:
    repo_dir = build_temp_repo()
    result = run_act_push(repo_dir)
    combined_output = result.stdout + "\n" + result.stderr

    with open(ACT_RESULT_PATH, "a", encoding="utf-8") as f:
        f.write("=" * 78 + "\n")
        f.write("TEST RUN: pr-label-assigner workflow via `act push --rm`\n")
        f.write(f"Exit code: {result.returncode}\n")
        f.write("=" * 78 + "\n")
        f.write(combined_output)
        f.write("\n")

    shutil.rmtree(repo_dir, ignore_errors=True)

    failures = []

    if result.returncode != 0:
        failures.append(f"act push exited with {result.returncode}, expected 0")

    if EXPECTED_UNIT_TEST_SUMMARY not in combined_output:
        failures.append(f"Expected unit test summary '{EXPECTED_UNIT_TEST_SUMMARY}' not found in act output")

    for scenario, expected_line in EXPECTED_SCENARIO_OUTPUT.items():
        if expected_line not in combined_output:
            failures.append(f"Scenario '{scenario}': expected exact output '{expected_line}' not found")

    success_count = combined_output.count("Job succeeded")
    if success_count < 2:
        failures.append(f"Expected 'Job succeeded' to appear at least twice (one per job), found {success_count}")

    if failures:
        print("ACT TEST HARNESS FAILURES:", file=sys.stderr)
        for f_ in failures:
            print(f" - {f_}", file=sys.stderr)
        return 1

    print("All act-based assertions passed.")
    print(f"Full output appended to {ACT_RESULT_PATH}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: results/2026-07-01_184135/tasks/15-test-results-aggregator/typescript-bun-fable5-high/generated-code/scripts/run-act-tests.ts
/**
 * End-to-end pipeline test harness: runs every test case THROUGH the GitHub
 * Actions workflow via `act` (nektos/act), never by invoking the aggregator
 * directly.
 *
 * For each case it:
 *   1. Creates a temp git repo containing the project files plus that case's
 *      fixture files copied to `results/` (what a matrix build would produce),
 *   2. Runs `act push --rm` and captures all output,
 *   3. Appends the output to ./act-result.txt (clearly delimited),
 *   4. Asserts: act exited 0, both jobs report "Job succeeded", no job
 *      failed, and the logs contain the EXACT expected values for that
 *      case's input (AGGREGATE_RESULT counters, flaky test names, markdown
 *      rows, and the bun test tally).
 *
 * Run with:  bun run scripts/run-act-tests.ts
 */
import { cp, mkdir, mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";

const ROOT = join(import.meta.dir, "..");
const ACT_RESULT_FILE = join(process.cwd(), "act-result.txt");

/** Files/dirs that make up the project inside each temp repo. */
const PROJECT_FILES = [
  "package.json",
  "tsconfig.json",
  ".actrc",
  "src",
  "tests",
  "fixtures",
  ".github",
] as const;

interface ActCase {
  name: string;
  /** Fixture directory (relative to repo root) copied to `results/`. */
  fixtureDir: string;
  /** Every string must appear verbatim in the act output. */
  expectExact: string[];
}

// Known-good expected values, hand-computed from the fixture files.
const CASES: ActCase[] = [
  {
    name: "case1-mixed-matrix-with-flaky-test",
    fixtureDir: "fixtures/case1",
    expectExact: [
      // 10 tests: 6 passed, 2 failed, 2 skipped; 4.9s total; 1 flaky.
      "AGGREGATE_RESULT total=10 passed=6 failed=2 skipped=2 duration=4.90s flaky=1",
      "FLAKY_TEST auth > test_flaky_network",
      "| ✅ Passed | 6 |",
      "| **Total** | **10** |",
      "| `auth > test_flaky_network` | junit-shard-macos.xml | junit-shard-ubuntu.xml |",
      "| `api > test_delete` | json-shard-windows.json | 500 from server |",
      "**Total duration:** 4.90s across 3 runs",
      // The unit-test job must run the whole suite green inside the container.
      "41 pass",
      "0 fail",
    ],
  },
  {
    name: "case2-all-green-matrix",
    fixtureDir: "fixtures/case2",
    expectExact: [
      "AGGREGATE_RESULT total=5 passed=5 failed=0 skipped=0 duration=1.50s flaky=0",
      "✅ All 5 tests passed.",
      "No flaky tests detected.",
      "**Total duration:** 1.50s across 2 runs",
      "41 pass",
      "0 fail",
    ],
  },
];

function sh(cmd: string[], cwd: string): { exitCode: number; output: string } {
  const proc = Bun.spawnSync(cmd, { cwd, stdout: "pipe", stderr: "pipe" });
  return {
    exitCode: proc.exitCode,
    output: proc.stdout.toString() + proc.stderr.toString(),
  };
}

/** Fail loudly with context; the harness is a test, so any assert kills it. */
function assert(condition: boolean, message: string): void {
  if (!condition) {
    console.error(`\n❌ ASSERTION FAILED: ${message}`);
    process.exit(1);
  }
}

async function setupTempRepo(testCase: ActCase): Promise<string> {
  const dir = await mkdtemp(join(tmpdir(), `aggregator-act-${testCase.name}-`));
  for (const entry of PROJECT_FILES) {
    await cp(join(ROOT, entry), join(dir, entry), { recursive: true });
  }
  // The case's fixture files become the `results/` dir a matrix build would produce.
  await mkdir(join(dir, "results"), { recursive: true });
  await cp(join(ROOT, testCase.fixtureDir), join(dir, "results"), { recursive: true });

  for (const cmd of [
    ["git", "init", "-b", "main"],
    ["git", "config", "user.email", "harness@example.com"],
    ["git", "config", "user.name", "Act Harness"],
    ["git", "add", "-A"],
    ["git", "commit", "-m", `act harness: ${testCase.name}`],
  ]) {
    const { exitCode, output } = sh(cmd, dir);
    assert(exitCode === 0, `git setup failed (${cmd.join(" ")}): ${output}`);
  }
  return dir;
}

async function main(): Promise<void> {
  await Bun.write(ACT_RESULT_FILE, ""); // start fresh each harness run
  let combined = "";

  for (const testCase of CASES) {
    console.log(`\n=== Running ${testCase.name} through act ===`);
    const repo = await setupTempRepo(testCase);

    const { exitCode, output } = sh(["act", "push", "--rm", "--pull=false"], repo);

    combined +=
      `\n${"=".repeat(78)}\n` +
      `=== ACT TEST CASE: ${testCase.name} (fixtures: ${testCase.fixtureDir}) ===\n` +
      `=== act exit code: ${exitCode} ===\n` +
      `${"=".repeat(78)}\n` +
      output;
    await Bun.write(ACT_RESULT_FILE, combined);

    await rm(repo, { recursive: true, force: true });

    // 1. act itself must succeed.
    assert(exitCode === 0, `${testCase.name}: act exited with ${exitCode} (see act-result.txt)`);

    // 2. Both jobs (unit-tests, aggregate) must report success, none may fail.
    const succeeded = (output.match(/Job succeeded/g) ?? []).length;
    assert(succeeded >= 2, `${testCase.name}: expected 2 "Job succeeded" markers, saw ${succeeded}`);
    assert(!output.includes("Job failed"), `${testCase.name}: a job reported "Job failed"`);

    // 3. Exact expected values for this case's fixture input.
    for (const expected of testCase.expectExact) {
      assert(
        output.includes(expected),
        `${testCase.name}: output is missing exact expected string:\n  ${expected}`,
      );
    }
    console.log(`✅ ${testCase.name} passed (${succeeded} jobs succeeded)`);
  }

  console.log(`\nAll ${CASES.length} act test cases passed. Full logs: ${ACT_RESULT_FILE}`);
}

main().catch((err: unknown) => {
  console.error(`harness error: ${err instanceof Error ? (err.stack ?? err.message) : String(err)}`);
  process.exit(1);
});
